### PR TITLE
Hirak branch

### PR DESCRIPTION
**Pull request by hiraksarkar**
### Submission type

* [ ] This submission adds a new dataset
* [x] This submission adds a new method
* [ ] This submission adds a new metric
* [ ] This submission adds a new task
* [ ] This submission adds a new Docker image
* [ ] This submission fixes a bug (link to related issue: )
* [ ] This submission adds a new feature not listed above

### Testing

* [x] This submission was written on a forked copy of SingleCellOpenProblems
* [ ] GitHub Actions "Run Benchmark" tests are passing on this base branch of this pull request (include link to passed test: )
* [x] If this pull request is not ready for review (including passing the "Run Benchmark" tests), I will open this PR as a draft (click on the down arrow next to the "Create Pull Request" button)

### Submission guidelines

* [x] This submission follows the guidelines in our [Contributing](../blob/master/CONTRIBUTING.md) document
* [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change

### CODE REVIEW

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    # Learn from reference
+    X = adata_sc.X.toarray()
+    X_norm = X / X.sum(1)[:, np.newaxis]
+    X_scaled = StandardScaler(with_mean=False).fit_transform(X_norm)
+
+    model = NMF(n_components=factors, init="random", random_state=17)
+    Ha = model.fit_transform(X_scaled)
+    Wa = model.components_
+
+    cluster_df = pd.DataFrame(adata_sc.obs["label"], columns=["label"])
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:37]: this can be done simply with `adata_sc.obs[["label"]].copy()`

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    thresh_certainty = [0] * n_types
+    for c in range(n_types):
+        thresh_certainty[c] = np.max(pos_neg_dict[c][1])
+    # Evaluation ends here
+
+    # Start run on actual spatial data
+    X_sp = adata.X.toarray()
+    X_sp_norm = X_sp / X_sp.sum(1)[:, np.newaxis]
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:82]: same here for normalization

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+)
+def nmfreg(adata):
+    n_types = adata.obsm["proportions_true"].size[1]
+    adata_sc = adata.uns["sc_reference"].copy()
+    factors = 30
+    projectiob_type = "l2"
+
+    # Learn from reference
+    X = adata_sc.X.toarray()
+    X_norm = X / X.sum(1)[:, np.newaxis]
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:30]: if you are trying to normalize raw counts, there are utils methods for that (so that it's consisnte across tasks). For instance, in this case I think this is the recommended function:
```python
from ....tools.normalize import 
...
log_cpm(adata)
```

there are more in `tools.normalize` that could be useful (and prob important to test?)

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    code_version=check_version("NMF-reg"),
+)
+def nmfreg(adata):
+    n_types = adata.obsm["proportions_true"].size[1]
+    adata_sc = adata.uns["sc_reference"].copy()
+    factors = 30
+    projectiob_type = "l2"
+
+    # Learn from reference
+    X = adata_sc.X.toarray()
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:29]: just wondering if it is necessary for this to be sparse ? Prob yes if the `scikit-learn.NMF` doesn't support sparse, but would double check just in case.

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    X_norm = X / X.sum(1)[:, np.newaxis]
+    X_scaled = StandardScaler(with_mean=False).fit_transform(X_norm)
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:31]: also here, this is a normalization step that is for sure quite impactful for downstream result. Is this common for this method? If yes I would maybe add a comment or maybe even include it in the function name? e.g. 
```python
def nmfreg_cpm_standardscale():
	...
	return 
```

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    model = NMF(n_components=factors, init="random", random_state=17)
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:33]: that's a very good point re random seed, is there any best practice in the repo? will ask

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    ).T
+
+    factor_to_best_celltype = np.argmax(factor_to_cluster_map, axis=0)
+    celltype_to_best_factor = np.argmax(factor_to_cluster_map, axis=1)
+
+    factor_to_best_celltype_matrix = np.zeros((factors, n_types))
+    for i, j in enumerate(factor_to_best_celltype):
+        factor_to_best_celltype_matrix[i, j] = 1
+
+    Ha_norm = StandardScaler(with_mean=False).fit_transform(Ha)
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:57]: same here, just wondering again

File: openproblems/tasks/spatial_decomposition/methods/vanillanmf.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+@method(
+    method_name="Non-Negative Matrix Factorization (NMF).",
+    paper_name="Fast local algorithms for large scale nonnegative matrix and tensor factorizations",
+    paper_url="https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.214.6398&rep=rep1&type=pdf",
+    paper_year=2009,
+    code_url="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html",
+    code_version=check_version("scikit-learn"),
+)
+# Used from https://github.com/theislab/scanpy/issues/181#issuecomment-534867254
+def grouped_obs_mean(adata, group_key, layer=None):
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/vanillanmf.py:19]: I have the same function in my branch, when both are merged, let's settle on one and use that in all cases where means over cell types are needed.

File: openproblems/tasks/spatial_decomposition/methods/vanillanmf.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    )
+
+    # Make profiles from single-cell expression
+    # dataset
+    profile_mean = grouped_obs_mean(adata_sc, "celltype")
+    X = adata.X
+
+    Wa = vanila_nmf_model.fit_transform(X, H=profile_mean.values)
+
+    prop = Wa_norm = Wa / Wa.sum(1)[:, np.newaxis]
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/vanillanmf.py:64]: probably `Wa_norm` not needed ?

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+import pandas as pd
+import scipy.optimize
+import scipy.stats
+
+
+@method(
+    method_name="NMF-reg",
+    paper_name="Slide-seq: A scalable technology for measuring genome-wide expression at high spatial resolution",
+    paper_url="https://science.sciencemag.org/content/363/6434/1463",
+    paper_year=2009,
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:18]: don't think it's 2009 😅

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    code_url="N/A",
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:19]: you can just link aleks github repo I think ?

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
+from ....tools.decorators import method
+from ....tools.normalize import log_cpm
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/nmfreg.py:2]: weird, I would have expected that this unused import would have been catched by precommit... do you have flake8 enabled otherwise? pre-commit seems to pass anyway, not a problem, just owndering

File: openproblems/tasks/spatial_decomposition/methods/vanillanmf.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
+from ....tools.decorators import method
+from ....tools.normalize import log_cpm
```
Comment (giovp) [inline openproblems/tasks/spatial_decomposition/methods/vanillanmf.py:2]: same here

**Review by giovp — CHANGES_REQUESTED** [review]
main comments are regarding normalization steps (both cpm/logcpm and standard scale), if they are needed, they should be kept consistent to repo utils standards imho and maybe also clarify them in name ? meaning that there could be a `nmfreg_logcpm` and a `nmfreg_logcpm_standardscale` and they could have diff results? not sure not an expert

also the `obs_means` function I think is the same that we'll need for different implementations, so after this and my PR is merged, let's settle on one! (and test it!)

for the rest, looks awesome! great addition!

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py (hunk 1 above)
Comment (hiraksarkar) [reply openproblems/tasks/spatial_decomposition/methods/nmfreg.py:37]: Great catch!

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py (hunk 3 above)
Comment (hiraksarkar) [reply openproblems/tasks/spatial_decomposition/methods/nmfreg.py:30]: ok I saw the methods available, does not seem to do what we are trying to do here. Should I create a function in tools.normalize file @giovp? I guess this one is very method-specific.

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py (hunk 4 above)
Comment (hiraksarkar) [reply openproblems/tasks/spatial_decomposition/methods/nmfreg.py:29]: Good point, I converted to dense for later operations, like standard-scalar, normalization, row operations. NMF works with sparse.

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py (hunk 5 above)
Comment (hiraksarkar) [reply openproblems/tasks/spatial_decomposition/methods/nmfreg.py:31]: yes this makes sense, I am gonna just put that.

File: openproblems/tasks/spatial_decomposition/methods/nmfreg.py (hunk 3 above)
Comment (giovp) [reply openproblems/tasks/spatial_decomposition/methods/nmfreg.py:30]: I see ok, I think I'd just keep it as is then as indeed is method specific.
